cmd/pitr/recoverer/recoverer.go

```diff
@@ -0,0 +1,252 @@
+	}
+	err = r.GetBinlogList(startGTID)
+	if err != nil {
+		return errors.Wrap(err, "get binlog list")
+	}
+	err = r.Recover()
+	if err != nil {
+		return errors.Wrap(err, "recover")
+	}
+	return nil
+}
+
+func (r *Recoverer) DownloadBinlogs() error {
+	for _, binlog := range r.storage.ListObjects("") {
+		if strings.Contains(binlog, "set") {
+			continue
+		}
+		data, err := r.storage.GetObject(binlog)
+		if err != nil {
+			return errors.Wrap(err, "get object")
+		}
+		err = ioutil.WriteFile(binlog, data, 0644)
+		if err != nil {
+			return errors.Wrap(err, "write file")
+		}
+		r.binlogs = append(r.binlogs, binlog)
+	}
+	return nil
+}
+
+func (r *Recoverer) Recover() error {
+	host, err := r.getHost()
+	if err != nil {
+		return errors.Wrap(err, "get host")
+	}
+
+	for _, binlog := range r.binlogs {
+		err = r.DownloadBinlog(binlog)
+		if err != nil {
+			return errors.Wrap(err, "download binlog")
+		}
+		flags := ""
+
+		// TODO: add logic for all types
+		switch r.recoverType {
+		case Skip:
+			flags = "--exclude-gtids=" + r.excludingGTIDSet
+		case Transaction:
+		case Date:
+		case Latest:
+		}
+		cmdString := "mysqlbinlog /tmp/" + binlog + " " + flags + " | mysql -h" + host + " -u" + r.pxcUser + " -p$PXC_PASS"
+		cmd := exec.Command("sh", "-c", cmdString)
+		var outb, errb bytes.Buffer
+		cmd.Stdout = &outb
+		cmd.Stderr = &errb
+		err := cmd.Run()
+		if err != nil {
+			return errors.Wrap(err, "run cmd")
+		}
+		if errb.Bytes() != nil {
+			errors.Errorf("cmd error: %s", errb.String())
+		}
+		err = os.Remove("/tmp/" + binlog)
+		if err != nil {
+			return errors.Wrap(err, "remove binlog")
+		}
+	}
+
+	return nil
+}
+
+func (r *Recoverer) DownloadBinlog(binlog string) error {
+	content, err := r.storage.GetObject(binlog)
+	if err != nil {
+		return errors.Wrap(err, "get object")
+	}
+	f, err := os.Create("/tmp/" + binlog)
+	if err != nil {
+		return errors.Wrap(err, "create file")
+	}
+	defer f.Close()
+	_, err = f.Write(content)
+	if err != nil {
+		return errors.Wrap(err, "write content to file")
+	}
+	return nil
+}
+
+func (r *Recoverer) GetLastBackupGTID() (int64, error) {
+	sep := []byte("GTID of the last")
+
+	o, err := r.storage.GetObject(r.backupNAme + "/xtrabackup_info.lz4.00000000000000000000")
+	if err != nil {
+		return 0, errors.Wrap(err, "get object")
+	}
+
+	startIndex := bytes.Index(o, sep)
+	if startIndex == -1 {
+		return 0, errors.New("no gtid  data in backup")
+	}
+	newOut := o[startIndex+len(sep):]
+	e := bytes.Index(newOut, []byte("'\n"))
+	if e == -1 {
+		return 0, errors.New("cant find gtid  data in backup")
+	}
+
+	set := newOut[:e]
+	setArr := bytes.Split(set, []byte("-"))
+	if len(setArr) < 2 {
+		return 0, errors.New("cant find lastgtid in backup")
+	}
+	idBytes := setArr[len(setArr)-1]
+
+	id, err := strconv.ParseInt(string(idBytes), 10, 64)
+	if err != nil {
+		return 0, errors.New("cant convert last gtid to int64")
+	}
+
+	return id, nil
+}
+
+func (r *Recoverer) GetBinlogList(startID int64) error {
```

Comment:
"Get" method in fact "sets" - it's confusing. I reckon here is would be better to return a file list pass it to the next method rather than set inside the object. It would be more explicit. It is hard to observe functions' side effects what makes code harder to comprehend. Moreover returning a list you have some kind of atomicity - all or nothing. Setting the internal state incrementally you potentially expose intermediate (incomplete) results.

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/673/commits/74347e94f05b3b59a0399735a7a9e3ea4ae3d768